### PR TITLE
Fix bug caused by cached representation in cohomology classes

### LINKED ISSUE #35649: minimal_model of GradedCommutativeAlgebra misuses cache
### Is there an existing issue for this?

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.

### Did you read the documentation and troubleshoot guide?

- [X] I have read the documentation and troubleshoot guide

### Environment

```markdown
- **OS**: Debian Linux
- **Sage Version**: 9.8
```

### Steps To Reproduce

1. Create two GradedCommutativeDifferentialAlgebras with the same names in the generators but different minimal models
2. Compute the minimal model of the first one
3. Compute the minimal model of the second one

### Expected Behavior

The codomain of the minimal model of the second GCDA should be the second algebra itself

### Actual Behavior

It returns the first algebra

### Additional Information

Example:

```
sage: A.<e1,e2,e3,e4,e5,e6> = GradedCommutativeAlgebra(QQ)
sage: B1 = A.cdg_algebra({e5:e1*e2,e6:e3*e4})
sage: B2 = A.cdg_algebra({e5:e1*e2,e6:e1*e2+e3*e4})
sage: B1.minimal_model()
Commutative Differential Graded Algebra morphism:
  From: Commutative Differential Graded Algebra with generators ('x1_0', 'x1_1', 'x1_2', 'x1_3', 'y1_0', 'y1_1') in degrees (1, 1, 1, 1, 1, 1) over Rational Field with differential:
   x1_0 --> 0
   x1_1 --> 0
   x1_2 --> 0
   x1_3 --> 0
   y1_0 --> x1_0*x1_1
   y1_1 --> x1_2*x1_3
  To:   Commutative Differential Graded Algebra with generators ('e1', 'e2', 'e3', 'e4', 'e5', 'e6') in degrees (1, 1, 1, 1, 1, 1) over Rational Field with differential:
   e1 --> 0
   e2 --> 0
   e3 --> 0
   e4 --> 0
   e5 --> e1*e2
   e6 --> e3*e4
  Defn: (x1_0, x1_1, x1_2, x1_3, y1_0, y1_1) --> (e1, e2, e3, e4, e5, e6)
sage: B2.minimal_model()
Commutative Differential Graded Algebra morphism:
  From: Commutative Differential Graded Algebra with generators ('x1_0', 'x1_1', 'x1_2', 'x1_3', 'y1_0', 'y1_1') in degrees (1, 1, 1, 1, 1, 1) over Rational Field with differential:
   x1_0 --> 0
   x1_1 --> 0
   x1_2 --> 0
   x1_3 --> 0
   y1_0 --> x1_0*x1_1
   y1_1 --> x1_2*x1_3
  To:   Commutative Differential Graded Algebra with generators ('e1', 'e2', 'e3', 'e4', 'e5', 'e6') in degrees (1, 1, 1, 1, 1, 1) over Rational Field with differential:
   e1 --> 0
   e2 --> 0
   e3 --> 0
   e4 --> 0
   e5 --> e1*e2
   e6 --> e3*e4
  Defn: (x1_0, x1_1, x1_2, x1_3, y1_0, y1_1) --> (e1, e2, e3, e4, e5, -e5 + e6)
```

### PR DESCRIPTION
**Pull request by miguelmarco**
### :books: Description

Fixes #35649

Adds the parent to the creation of cohomology classes, so they are not confused between different parents.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/algebras/commutative_dga.py [hunk 1]
```diff
@@ -3893,10 +3893,54 @@ class CohomologyClass(SageObject, CachedRepresentation):
...
+           e2 --> 0
+           e3 --> 0
+           e4 --> 0
+           e5 --> e1*e2
+           e6 --> e1*e2 + e3*e4
+          Defn: (x1_0, x1_1, x1_2, x1_3, y1_0, y1_1) --> (e1, e2, e3, e4, e5, -e5 + e6)
+
     """
-    def __init__(self, x):
+    def __init__(self, x, parent = None):
```
Comment (mkoeppe) [inline src/sage/algebras/commutative_dga.py]: No spaces around `=` in parameters

Otherwise LGTM

Comment (miguelmarco) [reply src/sage/algebras/commutative_dga.py]: Fixed

**Review by mkoeppe — APPROVED** [review]

**Comment by miguelmarco** [comment]
Should the issue be changed to positive review too?

**Comment by mkoeppe** [comment]
No need for a status label on the issue; I've removed it.
It will be closed as completed automatically when the PR is merged.

**Comment by tscrim** [comment]
Minor point, but I don't think we should call the input `parent` as it is not the parent (which would be the not-(yet-)coded cohomology ring). Instead it should be something like `cdga`. That way it will also be future compatible.

File: src/sage/algebras/commutative_dga.py [hunk 2]
```diff
@@ -3893,10 +3893,54 @@ class CohomologyClass(SageObject, CachedRepresentation):
...
-        sage: CohomologyClass(x^2+2*y*z)
+        sage: CohomologyClass(x^2+2*y*z, A)
         [2*y*z + x^2]
+
+    In order for the cache to not confuse objects with the same representation,
+    we can pass the parent of the representative as a parameter.
+
+    TESTS::
+
+        sage: A.<e1,e2,e3,e4,e5,e6> = GradedCommutativeAlgebra(QQ)
```
Comment (tscrim) [inline src/sage/algebras/commutative_dga.py]:
```suggestion

    TESTS:
    
    In order for the cache to not confuse objects with the same representation,
    we can pass the parent of the representative as a parameter::

        sage: A.<e1,e2,e3,e4,e5,e6> = GradedCommutativeAlgebra(QQ)
```
Sorry, one last thing. This way the comment is seen as part of the TEST block and not a dangling (technical implementation detail) comment.

**Review by tscrim — APPROVED** [review]
Thank you. LGTM.

**Comment by dimpase** [comment]
b3d322ee8e98eb666ddc18fe9d94973202a70b4a
added extra trailing space - how didn't you see this on the linters?

**Comment by dimpase** [comment]
Perhaps the linter only shows the 1st trailing whitespace it finds, and so it didn't show this changed file

**Pull request merged**